nixos/lib/testing/matrix.nix

```
@@ -0,0 +1,255 @@
+          An attribute set where each attribute name corresponds to a choice; a single outcome of a decision.
+
+          When you define `matrix.foo.choice.bar`, it creates a copy of the test configuration inside `matrix.foo.choice.bar.module`, where the module argument `foo` will be set to `bar`.
+
+          `runTest` will traverse all matrix choices, so that in all the final test configurations it returns, a choice for `foo` has been made.
+        '';
+      };
+      after = mkOption {
+        description = mdDoc ''
+          This option affects the order in which `runTest` decides which
+          decisions to make first.
+
+          Instead of reading for example
+          `matrix.bar.choice.qux.module.matrix.foo.choice.xyz`,
+          you can define
+
+          ```
+            matrix.bar.after = ["foo"];
+          ```
+
+          so that `runTest` will read
+          `matrix.foo.choice.xyz.module.matrix.bar.choice.qux` instead.
+
+          This would usually amount to the same thing, unless you want the set
+          of `choice` attributes to be affected by the value of `foo`, which
+          isn't possible in the default ordering.
+
+          You'd typically use this with [`matrix.<name>.enable`](#opt-matrix._name_.enable)
+          to make sure that definitions from "earlier" decisions are taken into account.
+        '';
+        type = types.listOf types.str;
+        default = [ ];
+      };
+      enable = mkOption {
+        type = types.bool;
+        default = true;
+        description = mdDoc ''
+          Allows a decision to be disabled. When set to `false`, none of the
+          choices will be applied, and the depth of the relevant subtree
+          returned by `runTest` will be reduced by one node.
+
+          By avoiding useless test parameter combinations, you can reduce your testing workload,
+          and by avoiding nonsensical combinations, you can make all reasonable tests pass.
+
+          Make sure that `runTest` doesn't inspect this
+          decision too soon, thereby ensuring that all required definitions
+          are available. A typical use may look as follows:
+
+          ```nix
+          # For this example, we suppose that our application has a backend
+          # selection option, which only allowed when dynamic linking is enabled.
+
+          { linking, ... }:
+          {
+            matrix.backend = {
+              after = ["linking"];
+              enable = linking != "static";
+
+              # Neither choice module will be applied when static linking is enabled.
+              # If static linking does need some extra settings, set them in
+              # `matrix.linking.choice.static.module` instead.
+              choice.rest.module = ...;
+              choice.smtp.module = ...;
+            };
+          }
+          ```
+        '';
+      };
+    };
+  };
+
+  choiceModule = decisionName: extendModules: { name, ... }: {
+    options = {
```

Maybe

```suggestion
    options = {
      enable = ...
```

An easier way to filter out nonsensical/unsupported combinations.

---

Done.